### ISSUE TITLE
samtools mpileup: -l and -r give different results for same region

### ISSUE THREAD

**Issue by tseemann** [samtools/samtools#79]
I am getting a different mpileup depending on whether I query via -l (BED file) or -r (seq:begin-end). In particular, the bases after the indel have zero depth despite obvious read alignments.

<PRE>
% cat test.bed

Contig_Plasmid	599	605

% samtools mpileup aln.bam -f ref.fa -l test.bed | cut -c1-50

Contig_Plasmid	600	T	183	.$.$.$.$,.-1G.-1G,-1g,-1g
Contig_Plasmid	601	G	181	,************************
Contig_Plasmid	602	G	181	,..,,,.,,,,,,............
Contig_Plasmid	603	G	184	a..,,,.,,,,,,............
Contig_Plasmid	604	A	184	,$..,,,.,,,,,,...........
Contig_Plasmid	605	A	183	..,,,.,,,,,,............,

% samtools mpileup aln.bam -f ref.fa -r Contig_Plasmid:600-605 | cut -c1-50

Contig_Plasmid	600	T	183	.$.$.$.$,.-1G.-1G,-1g,-1g
Contig_Plasmid	601	G	0
Contig_Plasmid	602	G	0
Contig_Plasmid	603	G	3	^].^].^].	BCB
Contig_Plasmid	604	A	182	..,,,.,,,,,,............,
Contig_Plasmid	605	A	182	..,,,.,,,,,,............,
</PRE>

**Comment by kartong88** [comment]
It is an old problem based on Heng's reply on the mailing list. The cause for the difference is also explained on the page.

http://sourceforge.net/p/samtools/mailman/samtools-help/thread/443287A31897D446961707538481382A01765A@ICTS-S-MBX7.luna.kuleuven.be/

**Comment by kirkmcclure** [comment]
Whatever the history, a simple test case shows that -r does not currently handle insertions or deletions correctly.
````
> cat /tmp/issue79example.sam
@SQ     SN:c2   LN:9
s1      0       c2      1       0       4M      *       0       0       CTAA  XXXX
s1d1a   0       c2      1       0       1M1D2M  *       0       0       CAA   *
s1d1b   0       c2      1       0       2M1D1M  *       0       0       CTA   *
s1i1a   0       c2      1       0       1M1I3M  *       0       0       CGTAA *
s1i1b   0       c2      1       0       2M1I2M  *       0       0       CTGAA *
s1i1c   0       c2      1       0       3M1I1M  *       0       0       CTAGA *

> samtools view -b -o /tmp/issue79example.bam /tmp/issue79example.sam
> samtools index /tmp/issue79example.bam
> cat /tmp/issue79.fa
>c2
CTAATAATC

> samtools mpileup -f /tmp/issue79.fa /tmp/issue79example.sam
[mpileup] 1 samples in 1 input files
<mpileup> Set max per-file depth to 8000
c2      1       C       6       ^!.^!.-1T^!.^!.+1G^!.^!.        X~~~~~
c2      2       T       6       .*.-1A..+1G.    X~~~~~
c2      3       A       6       ..*...+1G       X~~~~~
c2      4       A       6       .$.$.$.$.$.$    X~~~~~

> samtools mpileup -f /tmp/issue79.fa /tmp/issue79example.bam
[mpileup] 1 samples in 1 input files
<mpileup> Set max per-file depth to 8000
c2      1       C       6       ^!.^!.-1T^!.^!.+1G^!.^!.        X~~~~~
c2      2       T       6       .*.-1A..+1G.    X~~~~~
c2      3       A       6       ..*...+1G       X~~~~~
c2      4       A       6       .$.$.$.$.$.$    X~~~~~
````
sensible so far, but:
````
> samtools mpileup -r c2 -f /tmp/issue79.fa /tmp/issue79example.bam
[mpileup] 1 samples in 1 input files
<mpileup> Set max per-file depth to 8000
c2      1       C       5       ^!.^!.^!.+1G^!.^!.      BB.=B
c2      2       T       5       ..-1A..+1G.     EE1=D
c2      3       A       4       ....+1G B111
c2      4       A       4       .$.$.$.$        @11.
````
is clearly wrong.
To complete the thought, -l is not obviously wrong:
````
> cat /tmp/issue79.bed
c2      0       4
> samtools mpileup -l /tmp/issue79.bed -f /tmp/issue79.fa /tmp/issue79example.bam
[mpileup] 1 samples in 1 input files
<mpileup> Set max per-file depth to 8000
c2      1       C       6       ^!.^!.-1T^!.^!.+1G^!.^!.        X~~~~~
c2      2       T       6       .*.-1A..+1G.    X~~~~~
c2      3       A       6       ..*...+1G       X~~~~~
c2      4       A       6       .$.$.$.$.$.$    X~~~~~
````

**Comment by mfcovington** [comment]
@kirkmcclure It's not a bug, it's a feature! :bug:

**Comment by jkbonfield** [comment]
It's related to the BAQ alignment code.  If you use -B to disable that then the results come back again just fine.

All this is further complicated by the -Q parameter which defaults to skipping bases below 13. See https://github.com/samtools/samtools/issues/385 for more on that.  However in this example even with -Q0 it is not having an effect.  Also none of this explains why filtering with -r different to no filtering at all.  It should have zero impact.

**Comment by jkbonfield** [comment]
So this boils down to undocumented "features" I think, which realistically are bugs but are a consequence of deliberate decisions.

In mpileup() func
```
    if (tid0 >= 0 && conf->fai) { // region is set
        ref = faidx_fetch_seq(conf->fai, h->target_name[tid0], 0, 0x7fffffff, &ref_len);
        ref_tid = tid0;
        for (i = 0; i < n; ++i) data[i]->ref = ref, data[i]->ref_id = tid0;
    } else ref_tid = -1, ref = 0;
```
tid0 comes from the iterator, which is only used when using -r.  This ref=0 unless using -r, even when we have a .fa, .fa.idx and (obviously) there are references in our input SAM/BAM file.

In mplp_func()
```
        has_ref = (ma->ref && ma->ref_id == b->core.tid)? 1 : 0;
        if (has_ref && (ma->conf->flag&MPLP_REALN)) bam_prob_realn_core(b, ma->ref, (ma->conf->flag & MPLP_REDO_BAQ)? 7 : 3);
```

Here we have to have ma->ref set, which will be zero from the above code unless using -r.  Thus bam_prob_realn_core can only be called when using -r.  There is no documentation to imply that BAQ only works with regions, but that's how it works.

Note that -l is different to -r, and also as far as I can see undocumented in how it works.  -r uses the index to look up a region.  It'll create an iterator and in doing so then set the ->ref field.  -l however doesn't use an index.  Instead it streams the entire file past, filtering out uninteresting regions on the fly.  All very unintuitive, but this subtle difference is understood and exploited by many users now.   As a consequence of that difference, regions in bed files don't get the iterator set and so don't get BAQ computation done on them.

The long and the short - BAQ is the cause of the difference, and BAQ only applies with -r.  At the very least that needs documenting.

**Comment by jmarshall** [comment]
@jkbonfield: There is also [faidx_fetch_seq() code further down](https://github.com/samtools/samtools/blob/develop/bam_plcmd.c#L463-L468) that reloads `->ref`/`->ref_id` when `bam_mplp_auto()` moves from one chromosome to the next.  I haven't examined it while running, but are you quite sure that this won't set `ma->ref` even when not using `-r`?

**Comment by jkbonfield** [comment]
I put a breakpoint on bam_prob_realn_core and it's definitely not called.   However this test only has a single reference so maybe it's not triggering.

I also don't understand the logic of calling bam_prob_realn_core only if ma->ref_id == b->core.tid.  How can it differ?  If we're using sorted data, which we must, then surely ma->ref_id always matches b->core.tid.    Maybe the conditional is there to fix a crash, but in the wrong place - avoiding it instead of fixing why ref_id is out of sync with core.tid.

It does appear though that this is a deliberate thing - to only use BAQ when we have an iterator. I just can't work out why.

**Comment by jkbonfield** [comment]
Testing a bam file with c1, 2 and c3, mplp_func seems to always be called with one reference behind, so it always fails the ma->ref && ma->ref_id == b->core.tid check.

I suspect this is because it's not doing the load and cache ref bit in mplp_func but in the calling code _after_ the bam_mplp_auto loop.  So it's starting to look more like it's bugged afterall than deliberately avoiding BAQ alignment on non-region data. (Although that initial assignment of ref is clearly very carefully coded to only do it when a region is specified.)

Quite impenetrable!

**Comment by jkbonfield** [comment]
I have a solution in the works.

**Comment by tseemann** [comment]
Thanks James. 
SAMtools is a complicated legacy-burdened beast, your effort is appreciated!

**Comment by kirkmcclure** [comment]
Hello James,

Can you share a test case for #125 mpileup -C requires at least 3 readings ?

(It seems related to #79 and I've spent quite a bit of time casting about in the dark, but haven't found the symptom.)

Thanks, and cheers, Kirk

**Comment by jkbonfield** [comment]
Hmm that was ages ago so almost certainly no!  I'll have a quick look and see if it's easy to reproduce.  Possibly the #407 fixes the -E issue too.  It'd be great if so.

**Comment by jkbonfield** [comment]
I can confirm that #407 fixes both this bug and #125.
